### PR TITLE
openvpn: add easy-rss option

### PR DESCRIPTION
**Pull request by DomT4**
Here’s a dirty chunk of code as a proposal to resolve #25021. This can probably be made a lot prettier, which I’m open to.

### CODE REVIEW

**Comment by tdsmith** [comment]
Where do e.g. crt.key and certs_by_serial end up? I could be wrong but I thiiiiiiink they're placed relative to the scripts and not relative to the current working directory. Seems like they should land somewhere like etc or var that doesn't get smashed on upgrade.

Is there an advantage to bundling this with openvpn instead of packaging it separately?

File: Library/Formula/openvpn.rb [hunk 1]
```diff
@@ -11,11 +9,46 @@ class Openvpn < Formula
...
+        bins = Dir["#{libexec}/share/easy-rsa/*"].select { |f| File.executable? f }
+
+        Dir.glob(bins) do |f|
+          execname = File.basename(f, "")
+          (bin+execname).write <<-EOS.undent
+            #!/usr/bin/env bash
+            export EASY_RSA=#{libexec}/share/easy-rsa
+            #{f} "$@"
+          EOS
+          chmod 0755, bin+execname
```
Comment (tdsmith) [inline Library/Formula/openvpn.rb]: equivalently, `(bin/execname).write_env_script f, :EASY_RSA => libexec/"share/easy-rsa"`

File: Library/Formula/openvpn.rb [hunk 2]
```diff
@@ -11,11 +9,46 @@ class Openvpn < Formula
...
   def install
+    if build.with? "easy-rsa"
+      resource("easy_rsa").stage do
+        system "autoreconf", "-fvi"
+        system "./configure", "--prefix=#{libexec}", "--disable-silent-rules"
+        system "make", "install"
+
+        bin.mkpath
+        File.rename libexec/"share/easy-rsa/vars", libexec/"share/easy-rsa/vars.sample"
+        bins = Dir["#{libexec}/share/easy-rsa/*"].select { |f| File.executable? f }
```
Comment (tdsmith) [inline Library/Formula/openvpn.rb]: could do something like `(libexec/"share/easy-rsa").children.select { |f| f.executable? }` to get a list of `Pathname` objects, which makes life a little easier

File: Library/Formula/openvpn.rb [hunk 3]
```diff
@@ -11,11 +9,46 @@ class Openvpn < Formula
...
+
+        Dir.glob(bins) do |f|
```
Comment (tdsmith) [inline Library/Formula/openvpn.rb]: Is this the same as `bins.each do |f|`?

File: Library/Formula/openvpn.rb [hunk 4]
```diff
@@ -11,11 +9,46 @@ class Openvpn < Formula
...
+
   def install
+    if build.with? "easy-rsa"
+      resource("easy_rsa").stage do
+        system "autoreconf", "-fvi"
+        system "./configure", "--prefix=#{libexec}", "--disable-silent-rules"
+        system "make", "install"
+
+        bin.mkpath
+        File.rename libexec/"share/easy-rsa/vars", libexec/"share/easy-rsa/vars.sample"
```
Comment (tdsmith) [inline Library/Formula/openvpn.rb]: you can write `mv` instead of `File.rename`

**Comment by DomT4** [comment]
> Seems like they should land somewhere like etc or var that doesn't get smashed on upgrade.

You're instructed to ` source ` wherever you've put them every time you run one of the binaries. It's designed to be portable in the extent that you can put them anywhere as long as you source it, but for the sake of following their recommendations I will probably install the ` vars.sample ` file into ` etc/"easy_rsa" ` directory when I fix up the rest of your recommendations here shortly.

> Is there an advantage to bundling this with openvpn instead of packaging it separately?

Until the version prior to this one OpenVPN shipped ` easy_rsa ` internally with ` OpenVPN `. It seemed reasonable to maintain that behaviour. Adam also suggested the resource mechanism in #25021.

File: Library/Formula/openvpn.rb (hunk 1 above)
Comment (DomT4) [inline Library/Formula/openvpn.rb]: Yeah, that should work, thanks. Will that take care of the ` chmod ` as well or?

File: Library/Formula/openvpn.rb (hunk 3 above)
Comment (DomT4) [inline Library/Formula/openvpn.rb]: Pretty much. I should have used ` bins.each do ` really. Will fix.

File: Library/Formula/openvpn.rb (hunk 4 above)
Comment (DomT4) [inline Library/Formula/openvpn.rb]: I didn't know if one was more Rubyesque than the other :smile_cat:

**Comment by DomT4** [comment]
Fixed up with a bunch of Tim's suggestions. Looks a chunk tidier now.

File: Library/Formula/openvpn.rb (hunk 1 above)
Comment (tdsmith) [inline Library/Formula/openvpn.rb]: Yep, oughta.

File: Library/Formula/openvpn.rb [hunk 5]
```diff
@@ -11,11 +9,43 @@ class Openvpn < Formula
...
+        system "./configure", "--prefix=#{libexec}", "--disable-silent-rules"
+        system "make", "install"
+
+        bin.mkpath
+        mkdir_p etc/"easy-rsa"
+        mv libexec/"share/easy-rsa/vars", etc/"easy-rsa/vars.sample"
+        bins = (libexec/"share/easy-rsa").children.select { |f| f.executable? }
+
+        bins.each do |f|
+          execname = File.basename(f, "")
```
Comment (tdsmith) [inline Library/Formula/openvpn.rb]: or `f.basename`, now that you have a `Pathname`

**Comment by tdsmith** [comment]
It looks like some of these, especially `EASYRSA_PKI`, can be set (in `vars` I guess) to change where the bookkeeping is kept: https://github.com/OpenVPN/easy-rsa/blob/master/doc/EasyRSA-Advanced.md#environmental-variables-reference. I think we'd still need to env-script it to help it find vars. Anything Homebrew-specific probably deserves a caveat.

> You're instructed to source wherever you've put them every time you run one of the binaries

I think I remember Ubuntu working this way but I don't see that in the instructions the project provides so I guess we have some freedom there?

> Until the version prior to this one OpenVPN shipped easy_rsa internally with OpenVPN. It seemed reasonable to maintain that behaviour.

The other argument is that if the OpenVPN developers decided to calve it off and package it separately, maybe we should take that as a signal to package it separately. The readme says it's a separate project and it seems like it ought to be useful independent of openvpn so I'm inclined to separate them.

**Comment by DomT4** [comment]
> The readme says it's a separate project and it seems like it ought to be useful independent of openvpn so I'm inclined to separate them.

If none of the other maintainers have any objections, I'm happy to close this and pen it as a new formula and continue the discussions on improvements there.

**Comment by DomT4** [comment]
Alright, I'm going to stick this on hold until v3.0.0 lands stable. Right now, ` easy-rsa ` expects all of the binaries, and the openssl.cnf files, and the vars file to be placed in the same directory. When that happens, everything is lovely. As soon as you try to script the binaries from elsewhere, or you try to ` export EASYRSA= ` the whole show goes to crap and failure galore is created. It expects, more or less, for someone to be ` cd `'ed into the ` EASYRSA ` directory, with all the conf files and var file present, and then for the user to do ` ./build-ca ` and similar - Which isn't really viable inside Homebrew.

So, slapping this on some cold ice until 3.0.0 lands stable. 3.0.0 replaces the 8+ binaries with 1 binary, 1 conf file, and an optional vars file, and generally makes life a less miserable place.

**Comment by tdsmith** [comment]
:+1:; didn't realize 3.x wasn't in release yet, sorry. Thanks for taking a look at this! Looking forward to being able to package it.

**Comment by DomT4** [comment]
Aye, still stuck on Release Candidate 2. Not sure what the hold up is upstream, but hopefully we'll get 3.x stable soon.

**Comment by acrogenesis** [comment]
soy no easy-rsa yet?

**Comment by DomT4** [comment]
Upstream still haven't released the 3.x branch stable, so not yet, alas.

**Comment by acrogenesis** [comment]
Sidenote, if anyone else needs easy-rsa know you can download tunnelblick and it comes with easy-rsa

**Comment by deisner** [comment]
Shouldn't this be "add easy-rsa option", not "easy-rss" ?

**Comment by DomT4** [comment]
Yes. There was a typo in the PR title. Not a big deal, given the quite dead status of this PR for now.

**Pull request closed without merging**